content/post/proteus.md

```diff
@@ -0,0 +1,117 @@
+
+We chose protocol buffers as the serialization format for this exchange of data. Usually, people create their `.proto` files where they define their models, enumerations, services and so on, and then the models in each specific language are generated from those `.proto` files.
+
+For us, there is a big problem with that approach: **code generated is not idiomatic** for any of the languages (take a look at the python code generated by any proto file and you will understand why).
+
+Instead of following this approach, we thought of something: what if we choose the language with the bigger role in our codebase and use it as the source of truth instead? That way, at least one of the language has idiomatic models and code. This language for us is Go.
+
+## Proteus
+
+Because of the aforementioned reason, we started working on [proteus](https://github.com/src-d/proteus), a tool that scans Go packages and generates `.proto` files from them.
+
+**How does it work?**
+
+* Scans all structs that have the comment `//proteus:generate` and generates them as protobuf messages.
+* Scans all the type definitions with the comment `//proteus:generate` and their constant values and transforms them into proper protobuf enumerations.
+* Resolves the types and ignores those which it can't.
+* Converts it to protobuf structures and writes the `.proto` files.
+
+So, imagine you have the following code:
+
+```go
+package models
+
+import "time"
+
+type Model struct {
+        ID        int64
+        CreatedAt time.Time
+}
+
+type User struct {
+        Model
+        Status             Status
+        Username           string
+        Password           string
+        NotCryptedPassword string `proteus:"-"`
+}
+
+type Status int
+
+const (
+        Pending  Status = iota
+        Active
+        Inactive
+)
+```
+
+That code will become the following protobuf file:
+
+```
+syntax = "proto3";
+package models;
+
+import "google/protobuf/timestamp.proto";
+
+message User {
+        int64 id = 1;
+        google.protobuf.Timestamp created_at = 2;
+        models.Status status = 3;
+        string username = 4;
+        string password = 5;
+}
+
+enum Status {
+        PENDING = 0;
+        ACTIVE = 1;
+        INACTIVE = 2;
+}
+```
+
+You can check for a more detailed example in the [examples folder](https://github.com/src-d/proteus/tree/master/example).
+
+You might wonder why we did not use [gengo](https://github.com/kubernetes/gengo) for this task. The reason we did not use it, it is because of the plugin system we want to implement in the future. The plugin system is not implemented yet, we still don't know how it's going to look, but having control of our source code allows us to integrate it in every way we need.
```

Comment:
If you want to explain that you found more problems in Gengo, that motivated us to not using it, I think it could be useful to explain it deeper in a separated section. It could be maybe named _"Looking for a solution"_, where you could list our "must have" features that could not be found in another tools.